cds-snc/report-a-cybercrime
# Added a TOTP code to bypass submission per day restriction

Fixed #1882

This PR adds a TOTP (Google Authenticator, Authy, Microsoft Authenticator, etc) code as a "bypass" to allow us to access the app even when we hit the report limit for the day.

This could be used for devs to test the app even when we are closed for the day.

Instructions to test:

1.  Generate a secret using the generatesecret.js node app. This is a stand-alone script with it's own packages. So you will need to 'npm i' separate from the app.

```
cd /utils/totp
npm i
node generatesecret.js

```
This will create two files. 
'secret' is a text file with the TOTP secret. 
'code.html' is a HTML file with a QR code of the secret

2. Copy the contents of /utils/totp/secret file and put it in your .env file:
Example:
TOTP_SECRET=OBKVGJLUHFEXQZZDMFVTWQTDFI2D4N3QEMSEM2ZRMRXGOWSUOVRA

3. Using an authenticator app of your choice, open the code.html file and scan the QR code.

4. Set your "SUBMISSIONS_PER_DAY" value very low in the .env (1)

5. Build and run the app. Don't forget to also "npm i" on the app to install new library (speakeasy)

6. Fill out some test reports. You need to submit more reports than your "SUBMISSIONS_PER_DAY" value. Once this is done, try to access the app and verify it redirects you to CAFC website

7. Once you cannot access your local app without being redirected, try the TOTP code to verify the bypass. Copy the 6 digit TOTP code from your authenticator and pass it to the app in the URL as a query parameter
Example:
http://localhost:3000/?totp=123456


Outstanding questions:
Right now the app works as normal once loaded. We could add code to let analysts know it's a test report and they can ignore it.
